### PR TITLE
Penetrating projectiles and resistance penetration 

### PR DESCRIPTION
**Pull request by Arendian**
## About the PR

This PR changes the resistance penetration system from a bool to a float, making it possible to give all damage specific resistance penetration values.
It also makes it possible for projectiles to penetrate entities based on the penetration power the projectile has:
Penetrating a wall uses twice as much penetration power compared to penetrating a mob(all entities should probably get their own penetration resistance in the future), all bullets can get additional or reduced penetration power from the gun that shoots it(currently only the sniper rifles do this).

- Normal bullet: No penetration by itself and loses 50% of it's damage every time it penetrates a living entity, is unable to penetrate walls.
- High velocity bullet: Penetrates 1 living entity by itself and loses 25% of it's damage every time it penetrates a mob, is unable to penetrate walls.
- AP bullet: Penetrates 2 living entities or 1 wall by itself and loses no damage when it penetrates an entity, ignores 50% of the targets resistances.
- Anti-Materiel bullet: penetrates 2 living entities or 1 wall by itself(snipers doubles these values) and loses no damage when it penetrates an entity, ignores 75% of the targets resistances, stamina damage removed.
- Red laser: Penetrate an infinite amount of living entities losing no damage and ignoring 30% of the target's resistances, unable to penetrate walls.
- X-ray: Penetrates all entities, including walls, without losing damage and ignores 30% of the target's resistances.
- Pulse: Does not penetrate and ignores 30% of the target's resistances.

Damage examples using a c20 and shooting a security officer wearing a bulletproof vest and helmet (0.4 x 0.9 = 0.36 multiplier)
Normal bullet: 100 / (16 x 0.36) = 17.36, 18 bullets to crit.
High velocity bullet: 100 / (19 x 0.36)  = 14.62, 15 bullets to crit.
AP bullet: 100 / (12 x ((1-0.36) x 0.5 + 0.36) = 12.25, 13 bullets to crit.

A few more examples against the same security officer:
Anti-Materiel bullet: 100 / (60 x ((1-0.36) x 0.75 + 0.36) = 50.4, 2 bullets to crit.
Normal .45 bullet: 100 / (35 x 0.36) = 7.93, 8 bullets to crit.
High velocity .45 bullet: 100 / (44 x 0.36) = 6.3, 7 bullets to crit.
AP .45 bullet: 100 / (28 x ((1-0.36) x 0.6 + 0.36) = 4.8, 5 bullets to crit.

Some pieces of armor have armor pierce reduction, reducing the effectiveness of armor piercing munitions: Deathsquad armor reduces the multiplier by 0.25 and Commander/Elite/ERT hardsuits reduce the penetration multiplier by 0.1.

Other changes:

- Guns now have an optional modifier that can increase or decrease the damage of the projectile it shoots.
- Blood-red hardsuit Pierce resist upped to 0.4 from 0.5, the PR that nerfed the helmets didn't compensate for the lost armor making someone in a blood-red hardsuit die from 13 .30 rifle bullets instead of the previous 17(25% effective health reduction).
At 0.4 someone in a bloodred with their helmet on has the same survivability again as a security officer wearing a bulletproof vest and a helmet(17 .30 rifle bullets).
- AP rounds for all guns printable at secfab, can be researched as a T2 tech.
- New shotgun shell type: High velocity slug.
- The advanced ammo bundle is added to the nuclear operative uplink costing 25 TC it contains magazines with advanced ammunition for the C20-r, L6 SAW and Bulldog (AP, HV, Incendiary, slug HV).

**Media**

https://github.com/space-wizards/space-station-14/assets/137322659/0ee90528-6a87-4aa6-a2cc-e78038715ab2

image
image

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl: Dygon

- add: Added AP rounds for every gun that ignore a percentage of their targets armor, obtainable from the secfab after a Tier 2 research.;
- add: Nuclear operatives can now buy the advanced ammo bundle containing HV, AP, Incendiary and the new HV slug rounds.
- tweak: All lasers now ignore 30% of the heat resistance of their target.
- tweak: Red Lasers and HV rounds now penetrate organic entities, hitting anyone behind the target.
- tweak: Anti-materiel rounds, AP Rounds and the X-ray cannon can now shoot through organic entities and walls.
- tweak: Anti materiel rounds lost their stamina damage in exchange for extra penetration power.

### CODE REVIEW

**Comment by metalgearsloth** [comment]
I don't think the order of magnitude of complexity added to penetration damage is worth it where sometimes it's 25% damage or 50% or some other value, especially if you need to write out a formula to explain it.

**Comment by EmoGarbage404** [comment]
i think that armor penetration as a whole should not be used in this granular way. If weapons are just going to ignore armor, it's going to become a balancing nightmare in terms of armor values vs. penetration values vs. raw damage. If armor for lasers is really an issue then it either needs to be reduced or lasers need to be buffed. Adding armor piercing to all of them is just goofy.

I also do not think that lasers or bullets should penetrate through multiple targets with the exception of the x-ray laser and possibly the sniper (given that it's balanced accordingly). There isn't a need for regular, HV, or AP rounds to hit multiple targets in a single shot.

**Comment by Arendian** [comment]
> I don't think the order of magnitude of complexity added to penetration damage is worth it where sometimes it's 25% damage or 50% or some other value, especially if you need to write out a formula to explain it.

The formula is reducing the armor of the target by the armor penetration power, so if someone has 50 % damage reduction, the armor penetrating bullet will reduce that to 25% damage reduction. only the anti materiel bullet and magnum have a different armor pen value because of their lower fire rate/higher caliber.

**Comment by metalgearsloth** [comment]
I think simpler would be much better:
1. Flat AP as before, either it pierces all armor or it doesn't.
2. No magic values for red lasers. Lasers themselves penetrating is problematic on a station because it's going to raycast across the entire thing and suck up a bunch of time + hit a bunch of random people up to 50 tiles away.
3. Just put flat penetration amounts on AP / HV rounds.

**Comment by Arendian** [comment]
> If armor for lasers is really an issue then it either needs to be reduced or lasers need to be buffed. Adding armor piercing to all of them is just goofy.

I added the laser armor penetration because another PR lowered all heat resistances in an attempt to make lasers more viable, but this introduced the issue of fire damage killing people wearing a supposed fireproof suit too fast.
At first I also only had X-ray penetrate all and the laser cannon penetrate mobs but the normal laser penetration was an attemp of me to make lasers more viable.


> I also do not think that lasers or bullets should penetrate through multiple targets with the exception of the x-ray laser and possibly the sniper (given that it's balanced accordingly). There isn't a need for regular, HV, or AP rounds to hit multiple targets in a single shot.
Would it be allright if they penetrated only if the target is wearing below x armor?

**Comment by Arendian** [comment]
> 3\. Just put flat penetration amounts on AP / HV rounds.

The game currently calculates every piece separately, so when I reduced it by a flat amount it would ignore all pieces of armor that had a lower resistance value than the penetration(all helmets/jumpsuits/weaker armor)

**Comment by LankLTE** [comment]
I don’t see why AP bullets need to ignore only half of their resistances. The formula I used for .45 was “20% less damage but ignores all armor” so it’s already only really effective against armor. Currently, if I’m understanding correctly, against a target with 30% armor which seems to be the standard now, it would ignore half of it (15%) and actually do less damage than a normal round still. I also intended to normalize HV ammo at a 25% damage increase but I’m not sure you messed with that so.. whatever.

**Comment by Arendian** [comment]
> I don’t see why AP bullets need to ignore only half of their resistances. The formula I used for .45 was “20% less damage but ignores all armor” so it’s already only really effective against armor. Currently, if I’m understanding correctly, against a target with 30% armor which seems to be the standard now, it would ignore half of it (15%) and actually do less damage than a normal round still. I also intended to standardize HV ammo at a 25% damage increase but I’m not sure you messed with HV damage.

The thought behind it was that using AP rounds on low armoured people should be punished(bullet flies through causing less damage)  while using it for their intended target(heavily armored people) would be an improvement, Also .45 has 60% armor Penetration and anti-materiel has 75% all other bullets have 50%. 
A deathsquad officer being killed in 3 seconds by a detective with AP rounds is a bit silly in my eyes.
But since I will need to overhaul it the numbers in the post shouldn't be considered anymore.

**Comment by LankLTE** [comment]
> A deathsquad officer being killed in 3 seconds by a detective with AP rounds is a bit silly in my eyes. 

There’s a reason I had restricted it to Syndies. I had considered making it a research but honestly AP ammo would be rare enough (I almost never see HV ammo used) for it not to need a nerf. 

> while using it for their intended target(heavily armored people) would be an improvement

As is they need over 40% armor for it to be an improvement, and even then Nukie hardsuits (the best “heavy armor” crew will be commonly facing) will only give a 5% damage increase which is not only basically negligible, but also worse than HV ammo. Even crew only have 60% armor at the most, against which AP ammo is _still_ worse than most HV Ammo. 

Also having different guns pretty arbitrarily have different Armor Piercing percentages is confusing for players.

**Comment by LankLTE** [comment]
Also personally I’d just give the bloodred helmet resistances again, 0.4 is pretty absurdly strong and they can still technically wear a helmet on top of that.

**Comment by Arendian** [comment]
> Also personally I’d just give the bloodred helmet resistances again, 0.4 is pretty strong and they can still technically wear a helmet on top of that.

Not a single helmet gives more than 10% pierce resist anymore(except deathsquad and swat) so them swapping helmets isn't going to help them, the 0.4 only brings their resistance back to where it was before the helmet armor got reduced.


> As is they need over 40% armor for it to be an improvement, and even then Nukie hardsuits (the best “heavy armor” crew will be commonly facing) will only give a 5% damage increase which is not only basically negligible, but also worse than HV ammo. Even crew only have 60% armor at the most, against which AP ammo is _still_ worse than most HV Ammo.

With the current numbers someone in a bloodred with their helmet on(0.4 x 0.9 = 0.36) dies in 5 hits from a .45 AP round and 7 hits from a .45 high velocity round. Also 13 rounds of .35 auto compared to 15 .35 high velocity.

But with maintainers wanting AP to completely ignore armor again I should probably scrap them since automatic AP rounds will be either too strong or too weak.

**Comment by GoodWheatley** [comment]
> i think that armor penetration as a whole should not be used in this granular way. If weapons are just going to ignore armor, it's going to become a balancing nightmare in terms of armor values vs. penetration values vs. raw damage. 

The granularity is useful though; right now how our AP system works is like having an armor system that either doesn't block any damage or completely negates it. With granular armor penetration it's easier to balance different calibers of bullets to have different effects, and lets niches be created between ammunition types. For example, this system actually makes automatic AP rounds possible to balance since you can adjust them to not completely ignore armor.

**Comment by GoodWheatley** [comment]
> > i think that armor penetration as a whole should not be used in this granular way. If weapons are just going to ignore armor, it's going to become a balancing nightmare in terms of armor values vs. penetration values vs. raw damage.
> 
> The granularity is useful though; right now how our AP system works is like having an armor system that either doesn't block any damage or completely negates it. With granular armor penetration it's easier to balance different calibers of bullets to have different effects, and lets niches be created between ammunition types. For example, this system actually makes automatic AP rounds possible to balance since you can adjust them to not completely ignore armor.

**Comment by metalgearsloth** [comment]
> For example, this system actually makes automatic AP rounds possible to balance since you can adjust them to not completely ignore armor.

Added complexity leads to clear best options going unnoticed and balance overall becoming worse, e.g. the cobra being the best traitor gun, where you start playing whack-a-mole with individual balancing.

**Comment by GoodWheatley** [comment]
> > For example, this system actually makes automatic AP rounds possible to balance since you can adjust them to not completely ignore armor.
> 
> Added complexity leads to clear best options going unnoticed and balance overall becoming worse, e.g. the cobra being the best traitor gun, where you start playing whack-a-mole with individual balancing.

A) the cobra is definitely not the best traitor gun
B) if there’s no clear best option, doesn’t that mean it’s well balanced?

**Comment by metalgearsloth** [comment]
> A) the cobra is definitely not the best traitor gun

It's certainly not worse than the revolver or the other pistol.

> B) if there’s no clear best option, doesn’t that mean it’s well balanced?

Which isn't the case right now with fewer levers to use.

**Comment by GoodWheatley** [comment]
> > B) if there’s no clear best option, doesn’t that mean it’s well balanced?
> 
> Which isn't the case right now with fewer levers to use.

so why are you arguing against adding an additional lever for adjusting armor penetration? I'm not saying the current changes in the PR are balanced, but rather that the functionalities it adds are useful for gun balance and gameplay in the future.

**Comment by Ilya246** [comment]
i have my own suggestion as to how AP should work, make it so (pseudocode) `damageTaken = bulletDamage * (1 - pow(1 - armorDamageMultiplier, APValue))`, this makes AP not completely obliterate armor multipliers like 0.01, unless you set the AP value high enough

**Comment by metalgearsloth** [comment]
On discord we bikeshedded reverting AP and just going with the per-gun damage mods instead.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
tl;dr

Keep per gun damage mods
Drop laser piercing
Dump AP changes as we'll try just gun damage mods.

**Comment by metalgearsloth** [comment]
Re-open if you come back to it.
